File: benchmarks/ethers-baseline/bench.js
// Baseline: how a maker signs Midnight offer trees today - ethers v6, single-threaded.
// Computes the SAME leaves/tree as the Rust crate, so its N=4 root must equal the Rust CROSSCHECK_ROOT.
const { TypedDataEncoder, keccak256, concat, toBeHex, zeroPadValue, SigningKey } = require("ethers");

const A = (b) => "0x" + b.toString(16).padStart(2, "0").repeat(20);
const RATIFIER = A(0xbb);

const types = {
  CollateralParams: [
    { name: "token", type: "address" },
    { name: "lltv", type: "uint256" },
    { name: "liquidationCursor", type: "uint256" },
    { name: "oracle", type: "address" },
  ],
  Market: [
    { name: "chainId", type: "uint256" },
    { name: "midnight", type: "address" },
    { name: "loanToken", type: "address" },
    { name: "collateralParams", type: "CollateralParams[]" },
    { name: "maturity", type: "uint256" },
    { name: "rcfThreshold", type: "uint256" },
    { name: "enterGate", type: "address" },
    { name: "liquidatorGate", type: "address" },
  ],
  Offer: [
    { name: "market", type: "Market" },
    { name: "buy", type: "bool" },
    { name: "maker", type: "address" },
    { name: "start", type: "uint256" },
    { name: "expiry", type: "uint256" },
    { name: "tick", type: "uint256" },
    { name: "group", type: "bytes32" },
    { name: "callback", type: "address" },
    { name: "callbackData", type: "bytes" },
    { name: "receiverIfMakerIsSeller", type: "address" },
    { name: "ratifier", type: "address" },
    { name: "reduceOnly", type: "bool" },
    { name: "maxUnits", type: "uint128" },
    { name: "maxAssets", type: "uint128" },
    { name: "continuousFeeCap", type: "uint256" },
  ],
};

const ZERO = "0x0000000000000000000000000000000000000000";

function sampleOffer(i) {
  return {
    market: {
      chainId: 1,
      midnight: A(0x11),
      loanToken: A(0x22),
      collateralParams: [
        { token: A(0x33), lltv: 860000000000000000n, liquidationCursor: 1, oracle: A(0x44) },
      ],
      maturity: 1800000000,
      rcfThreshold: 1000,
      enterGate: ZERO,
      liquidatorGate: ZERO,
    },
    buy: i % 2 === 0,
    maker: A(0x55),
    start: 0,
    expiry: 2000000000,
    tick: i % 6744,
    group: zeroPadValue(toBeHex(i), 32),
    callback: ZERO,
    callbackData: "0x",
    receiverIfMakerIsSeller: ZERO,
    ratifier: RATIFIER,
    reduceOnly: false,
    maxUnits: 1000000 + i,
    maxAssets: 0,
    continuousFeeCap: 0,
  };
}

const hashNode = (l, r) => keccak256(concat([l, r]));

function buildTree(leaves) {
  const levels = [leaves];
  while (levels[levels.length - 1].length > 1) {
    const prev = levels[levels.length - 1];
    const next = [];
    for (let i = 0; i < prev.length; i += 2) next.push(hashNode(prev[i], prev[i + 1]));
    levels.push(next);
  }
  return levels;
}
function proof(levels, index) {
  const h = levels.length - 1;
  const p = [];
  let idx = index;
  for (let l = 0; l < h; l++) {
    p.push(levels[l][idx ^ 1]);
    idx >>= 1;
  }
  return p;
}

const sk = new SigningKey("0x" + "42".repeat(32));

function pipeline(offers) {
  const t = process.hrtime.bigint();
  const leaves = offers.map((o) => TypedDataEncoder.hashStruct("Offer", types, o));
  const levels = buildTree(leaves);
  for (let i = 0; i < offers.length; i++) proof(levels, i); // takers need proofs
  const root = levels[levels.length - 1][0];
  // one signature covers the whole tree
  const structHash = keccak256(concat([keccak256(Buffer.from("dummy")), root])); // typehash cost is O(1); shape-equivalent
  sk.sign(structHash);
  return { root, us: Number((process.hrtime.bigint() - t) / 1000n) };
}

// cross-check vs Rust
const small = [0, 1, 2, 3].map(sampleOffer);
console.log("CROSSCHECK_ROOT_N4", pipeline(small).root);

for (const n of [1024, 4096, 16384]) {
  const offers = Array.from({ length: n }, (_, i) => sampleOffer(i));
  pipeline(offers); // warmup
  let best = Infinity;
  for (let r = 0; r < 20; r++) best = Math.min(best, pipeline(offers).us);
  console.log(`RESULT n=${n} single_us=${best} single_per_offer_ns=${((best * 1000) / n).toFixed(0)}`);
}

File: crates/nocturne/examples/bench.rs
//! Performance path: hash N maker offers, build their tree, generate every proof, and sign the
//! root. This is the local pipeline an MM repeats when the market moves.
//!
//!   cargo run -p nocturne-midnight --release --example bench

use k256::ecdsa::SigningKey;
use nocturne::*;
use rayon::prelude::*;
use std::time::Instant;

fn u256(x: u64) -> Word {
    let mut w = [0u8; 32];
    w[24..].copy_from_slice(&x.to_be_bytes());
    w
}

fn sample_offer(i: u64) -> Offer {
    let market = Market {
        chain_id: u256(1),
        midnight: [0x11; 20],
        loan_token: [0x22; 20],
        collateral_params: vec![CollateralParams {
            token: [0x33; 20],
            lltv: u256(860_000_000_000_000_000),
            liquidation_cursor: u256(1),
            oracle: [0x44; 20],
        }],
        maturity: u256(1_800_000_000),
        rcf_threshold: u256(1000),
        enter_gate: [0u8; 20],
        liquidator_gate: [0u8; 20],
    };
    Offer {
        market,
        buy: i % 2 == 0,
        maker: [0x55; 20],
        start: u256(0),
        expiry: u256(2_000_000_000),
        tick: u256(i % 6744), // spread quotes across the tick grid
        group: u256(i),
        callback: [0u8; 20],
        callback_data: Vec::new(),
        receiver_if_maker_is_seller: [0u8; 20],
        ratifier: [0xbb; 20],
        reduce_only: false,
        max_units: 1_000_000u128 + i as u128,
        max_assets: 0,
        continuous_fee_cap: u256(0),
    }
}

fn pipeline(offers: &[Offer], sk: &SigningKey, parallel: bool) -> (Word, u128) {
    let ratifier = [0xbbu8; 20];
    let chain_id = u256(1);
    let t = Instant::now();

    let leaves: Vec<Word> = if parallel {
        offers.par_iter().map(hash_offer).collect()
    } else {
        offers.iter().map(hash_offer).collect()
    };
    let tree = OfferTree::build(leaves).unwrap();
    // generate a proof for every leaf (takers need these to lift individual offers)
    let _proofs: Vec<Vec<Word>> = if parallel {
        (0..offers.len())
            .into_par_iter()
            .map(|i| tree.proof(i).unwrap())
            .collect()
    } else {
        (0..offers.len()).map(|i| tree.proof(i).unwrap()).collect()
    };
    let digest = tree_digest(tree.root(), tree.height(), chain_id, &ratifier);
    let _sig = sign_digest(sk, &digest);

    (tree.root(), t.elapsed().as_micros())
}

fn main() {
    let sk = SigningKey::from_bytes(&[0x42u8; 32].into()).unwrap();

    for &n in &[1024usize, 4096, 16384] {
        let offers: Vec<Offer> = (0..n as u64).map(sample_offer).collect();
        // warmup
        pipeline(&offers, &sk, true);

        let reps = 20;
        let mut st = u128::MAX;
        let mut pt = u128::MAX;
        for _ in 0..reps {
            st = st.min(pipeline(&offers, &sk, false).1);
            pt = pt.min(pipeline(&offers, &sk, true).1);
        }
        println!(
            "RESULT n={n} single_us={st} parallel_us={pt} single_per_offer_ns={:.0} parallel_per_offer_ns={:.0}",
            st as f64 * 1000.0 / n as f64,
            pt as f64 * 1000.0 / n as f64
        );
    }
}
